src/Compilers/CSharp/Portable/BoundTree/BoundDecisionDag.cs

```diff
@@ -10,24 +10,27 @@
 
 namespace Microsoft.CodeAnalysis.CSharp
 {
-    partial class BoundDecisionDag
+    internal partial class BoundDecisionDag
     {
-        public IEnumerable<BoundDecisionDag> Successors()
+        private HashSet<LabelSymbol> _reachableLabels;
+        private ImmutableArray<BoundDecisionDagNode> _topologicallySortedNodes;
+
+        private static IEnumerable<BoundDecisionDagNode> Successors(BoundDecisionDagNode node)
         {
-            switch (this)
+            switch (node)
             {
-                case BoundEvaluationPoint p:
+                case BoundEvaluationDecisionDagNode p:
                     yield return p.Next;
                     yield break;
-                case BoundDecisionPoint p:
+                case BoundTestDecisionDagNode p:
                     Debug.Assert(p.WhenFalse != null);
                     yield return p.WhenFalse;
                     Debug.Assert(p.WhenTrue != null);
                     yield return p.WhenTrue;
                     yield break;
-                case BoundDecision d:
+                case BoundLeafDecisionDagNode d:
                     yield break;
-                case BoundWhenClause w:
+                case BoundWhenDecisionDagNode w:
                     Debug.Assert(w.WhenTrue != null);
                     yield return w.WhenTrue;
                     if (w.WhenFalse != null)
```

Comment:
Could as well return ImmutableArray?
```cs
return w.WhenFalse != null
  ? ImmurableArray.Create(w.WhenTrue, w.WhenFalse)
  : ImmutableArray.Create(w.WhenTrue);
```

Reply:
This is an iterator method. I think you're suggesting changing it to an ordinary method. I don't think it would reduce allocation (there is an array inside the `ImmutableArray`, and that immutable array would be boxed for the return value's `IEnumerable<...>`. In fact, I think that's *more* allocation.

---
In reply to: [177485292](https://github.com/dotnet/roslyn/pull/25739#discussion_r177485292) [](ancestors = 177485292)

Reply:
Thought that `if` is the only reason that this had to become an iterator *and* in turn, return IEnumerable. if the sort algorithm only works on an IEnumerable, yeah it won't make much of a difference, but we could possibly make all that ImmurableArray to avoid boxing all the way and simplify generated code.

Reply:
The generic topological sort is used in a few places, but yes it could probably be made to use an immutable array instead of an `IEnumerable`. I've recorded your suggestion as [https://github.com/dotnet/roslyn/issues/25754](https://github.com/dotnet/roslyn/issues/25754)

---
In reply to: [177498107](https://github.com/dotnet/roslyn/pull/25739#discussion_r177498107) [](ancestors = 177498107)